PCAP logging with fake TCP/UDP packet

sepalani opened the pull request:
This is more an experimental PR than anything else. It's faking Ethernet/IP headers using Dolphin's MAC and a dummy one. However, it lacks several things such as
- Proper TCP packets with flags, sequence number, etc.
- ~Dolphin's IP is hardcoded, FTM.~
- ~It's still WIP as I haven't tested many cases.~

What it does:
- Log SSL traffic
- Log TCP traffic
- Log UDP traffic
- Create a PCAP file (filename with gamename+timestamp) in Dump/SSL

Regardless, I tested it with different games and it seems to produce a valid PCAP file while logging TCP, UDP and SSL traffic. Config's ~~"NetworkDumpPCAP"~~ "DumpAsPCAP" can be switched to True/False to switch between binary SSL dump and TCP/UDP/SSL PCAP dump.

I'm quite skeptical regarding this method and really think a custom format (or another one) should probably be used instead within the PCAP file. Even though, after some testing it has interesting advantages such as allowing Wireshark to detect protocols such as SSL over 443 in TCP mode without using IOS's SSL built-ins and the use of other Wireshark features.

Anyway, ready to be reviewed & merged. Depends on https://github.com/dolphin-emu/dolphin/pull/5978.

iwubcode commented:
Is there a reason that we don't use a pcap library to handle the pcap output?  Or alternatively, is there a reason we don't just use something like tshark to capture on the particular ip/port?

Last question - what is this used for?

sepalani commented:
@iwubcode 
I don't really know why we don't use a pcap library for that. However, I know for a fact that is a really simple format. Then, IIRC, I think you need special privileges (can be tweeked using setcap) and install dependencies to use libpcap.

Regarding the tshark use, it was initially implemented by @delroth to reverse audio stuff, so completely different from network capture. It's used as a convenient format to dump data.

The SSL traffic's dumped using a binary/raw text format instead which isn't very suitable for binary protocol as we don't have any pieces of information concerning the sender and the receiver. That's why I tried the PCAP format and might use a custom format for each packet (if needed).

It could be used to dump data from a HLE point of view, that's to say, decrypted SSL data or anything providing the proper interface and HLE hooks.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+PCAPSSLCaptureLogger::PCAPSSLCaptureLogger()
+{
+  std::time_t t = std::time(nullptr);
+  char buffer[32] = "";
```
```cpp
std::ostringstream ss;
ss << std::put_time(std::localtime(&t), " %Y-%m-%d %Hh%Mm%Ss.pcap");

const std::string filename =
       File::GetUserPath(D_DUMPSSL_IDX) + SConfig::GetInstance().GetGameID() + ss.str();
...
```

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+std::vector<u8> PCAPSSLCaptureLogger::GetEthernetHeader(bool is_sender)
+{
+  constexpr std::size_t ETHERNET_HEADER_SIZE = 14;
+  std::vector<u8> ethernet_header(ETHERNET_HEADER_SIZE, 0);
```
`, 0` can be omitted. By default vectors zero out their contents on construction.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+                                                  u8 protocol, s32 socket)
+{
+  constexpr std::size_t IP_HEADER_SIZE = 20;
+  std::vector<u8> ip_header(IP_HEADER_SIZE, 0);
```
Ditto

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+void PCAPSSLCaptureLogger::LogTCP(bool is_sender, const void* data, std::size_t length, s32 socket)
+{
+  constexpr std::size_t TCP_HEADER_SIZE = 20;
+  std::vector<u8> tcp_header(TCP_HEADER_SIZE, 0);
```
Ditto

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+void PCAPSSLCaptureLogger::LogUDP(bool is_sender, const void* data, std::size_t length, s32 socket)
+{
+  constexpr std::size_t UDP_HEADER_SIZE = 8;
+  std::vector<u8> udp_header(UDP_HEADER_SIZE, 0);
```
Ditto

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+{
+  constexpr std::size_t ETHERNET_HEADER_SIZE = 14;
+  std::vector<u8> ethernet_header(ETHERNET_HEADER_SIZE, 0);
+  u8 mac[Common::MAC_ADDRESS_SIZE] = {};
```
`std::array`. This also allows replacing the memset below with either a `.fill(0)` or `mac = {};`.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+  std::vector<u8> tcp_header(TCP_HEADER_SIZE, 0);
+
+  // Get socket port
+  struct sockaddr_in sock_addr, peer_addr;
```
These should be on their own line.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+  tcp_header[15] = 0xFF;
+
+  // Add the packet
+  std::vector<u8> packet;
```
This can be moved closer to it's usage point.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+  udp_header[5] = udp_length & 0xFF;
+
+  // Add the packet
+  std::vector<u8> packet;
```
Ditto

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+    std::memset(mac, 0, Common::MAC_ADDRESS_SIZE);
+
+  // Write Dolphin MAC address, the other one is dummy
+  if (is_sender)
```
Comments should go on top of non-declarative code lines where applicable.

```cpp
if (is_sender)
{
    // Source MAC address
    std::copy(std::begin(mac), std::end(mac), ethernet_header.begin() + 6);
}
else
{
    // Destination MAC address
    std::copy(std::begin(mac), std::end(mac), ethernet_header.begin());
}

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,378 @@
+  std::vector<u8> udp_header(UDP_HEADER_SIZE, 0);
+
+  // Get socket port
+  struct sockaddr_in sock_addr, peer_addr;
```
These should be on their own line.

lioncash commented on Source/Core/Common/PcapFile.h:
```diff
@@ -22,9 +22,14 @@
 class PCAP final
 {
 public:
+  static constexpr u32 PCAP_DATA_LINK_TYPE = 147;  // Reserved for internal use.
```
You may want to place this into an enum class, along with the magic number used [here](#diff-2394256c451ee31017bb735ff6f6f762R94) to represent ethernet.

lioncash commented on Source/Core/Core/ConfigManager.h:
```diff
@@ -316,6 +318,8 @@ struct SConfig
   bool m_SSLVerifyCert;
   bool m_SSLDumpRootCA;
   bool m_SSLDumpPeerCert;
+  std::unique_ptr<Core::NetworkCaptureLogger> m_network_logger{
```
Should this not be lazy initialized? It seems kind of wasteful to allocate by default for this when the majority of users likely won't even need to use it.

sepalani commented:
@lioncash I think I addressed all your comments.

I tested it with Netflix and the result is not as bad as I expected. UDP based protocols seem fine such as DNS. TCP/SSL protocols even if they are a bit messed up, Wireshark doesn't have too much trouble handling them.

lioncash commented on Source/Core/Common/PcapFile.h:
```diff
@@ -22,9 +22,19 @@
 class PCAP final
 {
 public:
+  enum class Encapsulation : u32
```
Wouldn't this be more suited to be named LinkType or DataLinkType, since that's what field in the PCAP header it's used for/representing?

Also the general convention with enum classes in the codebase is to CamelCase them like we regularly do (including uppercasing initialisms): e.g.

```cpp
enum class LinkType : u32
{
    Ethernet,
    User
};
```

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,390 @@
+std::vector<u8> PCAPSSLCaptureLogger::GetEthernetHeader(bool is_sender)
+{
+  constexpr std::size_t ETHERNET_HEADER_SIZE = 14;
+  std::vector<u8> ethernet_header(ETHERNET_HEADER_SIZE);
```
Looking at this again, is there any reason std::vector is used, as opposed to std::array? Especially when it's fairly small and the size is known up front?

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,390 @@
+                                                  u8 protocol, s32 socket)
+{
+  constexpr std::size_t IP_HEADER_SIZE = 20;
+  std::vector<u8> ip_header(IP_HEADER_SIZE);
```
Ditto

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,390 @@
+void PCAPSSLCaptureLogger::LogTCP(bool is_sender, const void* data, std::size_t length, s32 socket)
+{
+  constexpr std::size_t TCP_HEADER_SIZE = 20;
+  std::vector<u8> tcp_header(TCP_HEADER_SIZE);
```
Ditto, wrt `std::array`.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,390 @@
+  std::vector<u8> ethernet_header = GetEthernetHeader(is_sender);
+  std::vector<u8> ip_header = GetIPHeader(is_sender, TCP_HEADER_SIZE + length, 0x06, socket);
+  std::vector<u8> packet;
+  packet.insert(packet.end(), std::make_move_iterator(ethernet_header.begin()),
```
`std::make_move_iterator` isn't needed here. There's no internal data within a `u8` that can be moved, it's just a primitive. There's no real difference between a move or a copy in this instance (either way you pass around the same amount of data).

`std::make_move_iterator` is beneficial when a type contains other data that can be moved out of it (for example, a `std::vector<std::string>` would be a good use-case, since the internal character buffer of each `std::string` instance be moved instead of being copied, saving on allocations).

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,390 @@
+  std::vector<u8> ethernet_header = GetEthernetHeader(is_sender);
+  std::vector<u8> ip_header = GetIPHeader(is_sender, UDP_HEADER_SIZE + length, 0x11, socket);
+  std::vector<u8> packet;
+  packet.insert(packet.end(), std::make_move_iterator(ethernet_header.begin()),
```
Ditto

lioncash commented on Source/Core/Core/NetworkCaptureLogger.h:
```diff
@@ -0,0 +1,107 @@
+
+#include "Common/CommonTypes.h"
+#include "Common/File.h"
+#include "Common/PcapFile.h"
```
Isn't `PCAP` able to be a forward declaration?

sepalani replied:
It doesn't seem to work on non-Windows builds:
> /usr/include/c++/5/bits/unique_ptr.h:74:22: error: invalid application of 'sizeof' to incomplete type 'PCAP'

lioncash replied:
This is because the containing class's (PCAPSSLCaptureLogger) destructor is implicitly defined, which makes it inlineable (inlined destructor + site of incomplete type -> boom). If you default the destructor in the cpp file, it should go away.

sepalani replied:
Done.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,390 @@
+    std::memcpy(&tcp_header[2], &peer_addr.sin_port, 2);
+
+    // Write sequence number
+    static u32 sequence_number = 0;
```
Is there any reason this and the other function-local mutable statics aren't class variables? This makes the class thread-unsafe in its current form.

lioncash commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,390 @@
+
+    // Write sequence number
+    static u32 sequence_number = 0;
+    tcp_header[4] = sequence_number >> 24 & 0xFF;
```
These assignments are truncating without a cast to u8 (datatype-wise). Implicit truncation is also present in other places of the class as well.

sepalani commented:
@lioncash Done.

sepalani commented:
I removed the "WIP" tag in the PR's name since I think the PR is ready to be reviewed & merged.

sepalani commented:
Rebased on latest master.

Still ready to be reviewed & merged.

sepalani commented:
Rebased on latest master.

sepalani commented:
Rebased and converted to draft.

I want to address the hardcoded IP issue and improve the UDP logging.

sepalani commented:
This PR is ready for review.

Regarding the original one, I moved some of the structures to Common/Network. I also changed the interface to take a `sockaddr` to log UDP properly.

iwubcode commented on Source/Core/Common/Network.h:
```diff
@@ -25,6 +27,78 @@ enum
+  u16 ethertype = 0;
+
+  EthernetHeader();
+  EthernetHeader(u16 ether_type);
```
`explicit`?

iwubcode commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -382,10 +383,42 @@ void PPCDebugInterface::RunToBreakpoint()
+    }
+    if (m_network_logger->GetCaptureType() != current_capture_type)
+      m_network_logger.reset();
+  } while (!m_network_logger);
```
I'm missing something, what's the point of this being a loop?

iwubcode commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,222 @@
+  return NetworkCaptureType::None;
+}
+
+void BinarySSLCaptureLogger::LogSSLRead(const void* data, std::size_t length, s32 socket)
```
Not a big deal but I wonder if we could combine this and writing into one function

iwubcode commented on Source/Core/Core/NetworkCaptureLogger.h:
```diff
@@ -0,0 +1,109 @@
+class NetworkCaptureLogger
+{
+public:
+  virtual ~NetworkCaptureLogger();
```
Rule of five?

iwubcode commented on Source/Core/Core/NetworkCaptureLogger.h:
```diff
@@ -0,0 +1,109 @@
+  virtual NetworkCaptureType GetCaptureType() const = 0;
+};
+
+class DummyNetworkCaptureLogger : public NetworkCaptureLogger
```
These loggers don't seem to be used anywhere?

But they could potentially be marked `final`

Rumi-Larry commented:
Just something that I'm curious of. Wouldn't merging this pr https://github.com/dolphin-emu/dolphin/pull/7817 make this code easier to mantain?

sepalani commented:
@Rumi-Larry No it won't help as this PR is unrelated to the Broadband adapter.

Ready to be reviewed & merged.

leoetlino commented on Source/Core/Core/Debugger/PPCDebugInterface.h:
```diff
@@ -82,9 +83,12 @@ class PPCDebugInterface final : public Common::DebugInterface
   u32 GetColor(u32 address) const override;
   std::string GetDescription(u32 address) const override;
 
+  std::shared_ptr<Core::NetworkCaptureLogger> NetworkLogger();
```
`#include <memory>`

leoetlino commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void BinarySSLCaptureLogger::LogRead(const void* data, std::size_t length)
+{
+  if (!Config::Get(Config::MAIN_NETWORK_SSL_DUMP_READ))
```
This check doesn't seem to be useful.

leoetlino commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void BinarySSLCaptureLogger::LogWrite(const void* data, std::size_t length)
+{
+  if (!Config::Get(Config::MAIN_NETWORK_SSL_DUMP_WRITE))
```
ditto

leoetlino commented on Source/Core/Common/Network.h:
```diff
@@ -25,6 +27,78 @@ enum
+  MACAddress source = {};
+  u16 ethertype = 0;
+
+  EthernetHeader();
```
Member functions should be above member variables

leoetlino commented on Source/Core/Common/Network.h:
```diff
@@ -25,6 +27,78 @@ enum
+  u8 ttl = 0;
+  u8 protocol = 0;
+  u16 header_checksum = 0;
+  u8 source_addr[IPV4_ADDR_LEN];
```
`u8 source_addr[IPV4_ADDR_LEN]{};`

leoetlino commented on Source/Core/Common/Network.h:
```diff
@@ -25,6 +27,78 @@ enum
+  u8 protocol = 0;
+  u16 header_checksum = 0;
+  u8 source_addr[IPV4_ADDR_LEN];
+  u8 destination_addr[IPV4_ADDR_LEN];
```
`u8 destination_addr[IPV4_ADDR_LEN];`

leoetlino commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -462,8 +462,10 @@ void WiiSocket::Update(bool read, bool write, bool except)
-                                                                 ret);
+              PowerPC::debug_interface.NetworkLogger()->LogSSLWrite(
+                  Memory::GetPointer(BufferOut2), ret,
+                  reinterpret_cast<mbedtls_net_context*>(Device::NetSSL::_SSL[sslID].ctx.p_bio)
```
A static_cast should be enough, I think?

leoetlino commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -495,7 +497,10 @@ void WiiSocket::Update(bool read, bool write, bool except)
-              PowerPC::debug_interface.NetworkLogger()->LogRead(Memory::GetPointer(BufferIn2), ret);
+              PowerPC::debug_interface.NetworkLogger()->LogSSLRead(
+                  Memory::GetPointer(BufferIn2), ret,
+                  reinterpret_cast<mbedtls_net_context*>(Device::NetSSL::_SSL[sslID].ctx.p_bio)
```
static_cast

leoetlino commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -553,10 +558,12 @@ void WiiSocket::Update(bool read, bool write, bool except)
-          const int ret = sendto(fd, data, BufferInSize, flags,
-                                 has_destaddr ? (struct sockaddr*)&local_name : nullptr,
-                                 has_destaddr ? sizeof(sockaddr) : 0);
+          sockaddr* to = has_destaddr ? (sockaddr*)&local_name : nullptr;
```
`auto* to = ... ? reinterpret_cast<...>(...) : nullptr;`

leoetlino commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -599,11 +606,13 @@ void WiiSocket::Update(bool read, bool write, bool except)
-          const int ret = recvfrom(fd, data, data_len, flags,
-                                   BufferOutSize2 ? (struct sockaddr*)&local_name : nullptr,
-                                   BufferOutSize2 ? &addrlen : nullptr);
+          sockaddr* from = BufferOutSize2 ? (sockaddr*)&local_name : nullptr;
```
See my previous comment.

leoetlino commented on Source/Core/Core/NetworkCaptureLogger.cpp:
```diff
@@ -49,4 +69,155 @@ NetworkCaptureType BinarySSLCaptureLogger::GetCaptureType() const
+
+PCAPSSLCaptureLogger::PCAPSSLCaptureLogger()
+{
+  std::time_t t = std::time(nullptr);
```
You can use `fmt` for this to write this in a more concise way, see https://fmt.dev/latest/api.html#date-and-time-formatting

sepalani commented:
@leoetlino
Done.

leoetlino commented on Source/Core/Common/Network.cpp:
```diff
@@ -72,4 +81,91 @@ std::optional<MACAddress> StringToMacAddress(std::string_view mac_string)
+      checksum += (word << 8) + b;
+    index += 1;
+  }
+  header_checksum = htons((checksum + (checksum >> 16) & 0xFFFF) ^ 0xFFFF);
```
Shouldn't this be a loop? Something like this:

```cpp
while (checksum > 0xFFFF)
  checksum = (checksum >> 16) + (checksum & 0xFFFF); 

header_checksum = htons(checksum ^ 0xFFFF);
```

If you only perform one iteration of the loop, you may still end up with a checksum that is larger than 0xFFFF.

sepalani replied:
Good catch. fixed.

leoetlino commented on Source/Core/Common/Network.cpp:
```diff
@@ -72,4 +81,91 @@ std::optional<MACAddress> StringToMacAddress(std::string_view mac_string)
+  std::memcpy(&destination_port, &to.sin_port, 2);
+  length = htons(Size() + data_length);
+}
+u16 UDPHeader::Size() const
```
Please add an empty line before this function.

leoetlino commented on Source/Core/Common/Network.cpp:
```diff
@@ -72,4 +81,91 @@ std::optional<MACAddress> StringToMacAddress(std::string_view mac_string)
+{
+  return static_cast<u16>(SIZE);
+}
+u8 UDPHeader::IPProto() const
```
here too

leoetlino commented on Source/Core/Common/Network.h:
```diff
@@ -25,6 +27,78 @@ enum
+  u8 source_addr[IPV4_ADDR_LEN]{};
+  u8 destination_addr[IPV4_ADDR_LEN]{};
+
+  IPv4Header();
```
This should be moved before the members as well.

leoetlino commented on Source/Core/Common/Network.h:
```diff
@@ -25,6 +27,78 @@ enum
+  u16 checksum = 0;
+  u16 urgent_pointer = 0;
+
+  TCPHeader();
```
ditto

leoetlino commented on Source/Core/Common/Network.h:
```diff
@@ -25,6 +27,78 @@ enum
+  u16 length = 0;
+  u16 checksum = 0;
+
+  UDPHeader();
```
ditto

leoetlino commented on Source/Core/Core/NetworkCaptureLogger.h:
```diff
@@ -0,0 +1,114 @@
+  enum class LogType
+  {
+    Read,
+    Write
```
```suggestion
    Write,
```

sepalani commented:
@leoetlino Done.

I also added the `ComputeNetworkChecksum` function to compute the IP checksum and also computed the TCP checksum value.

sepalani commented:
@dolphin-emu-bot rebuild
